rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,391 @@
+---
+authors: Alex McGrath (alex.mcgrath@goteleport.com)
+state: draft
+---
+
+# RFD 57 - Automatic discovery and enrollment of AWS servers
+
+## What
+
+Proposes a way by which an SSH service might automatically discover and register AWS
+EC2 instances.
+
+## Why
+
+Currently when adding a new AWS server, it's required that Teleport be installed
+after the server has been provisioned which may be a slow process for organizations
+with large numbers of servers as it needs to be installed and then added to the
+teleport cluster
+
+With the changes described in this document, Teleport will be able to resolve the
+issues with adding AWS servers to Teleport clusters automatically.
+
+
+## Discovery
+
+Discovery can use a matcher similar to the `db_service/aws` matcher, however EC2
+instances will have an optional install command and set of join parameters:
+
+```yaml
+ssh_service:
+  enabled: "yes"
+  aws:
+  aws:
+   - types: ["ec2"]
+     regions: ["eu-central-1"]
+     tags:
+       "teleport": "yes"
+     install:
+       join_params:
+         token_name:  aws-discovery-iam-token # default value
+     ssm:
+       document: "TeleportDiscoveryInstaller" # default value
+```
+
+The agent will use EC2's `DescribeInstances` API in order to list instances[1]. This
+will require the teleport SSH agent to include `ec2:DescribeInstances` as part of
+it's IAM permissions.
+
+As with AWS database discover, new EC2 nodes will be discovered periodically on a 60
+second timer, as new nodes are found they will be added to the teleport cluster.
+
+In order to avoid attempting to reinstall teleport on top of an instance where it is
+already present the generated teleport config will match against the node name using
+the aws account id and instance id.
+
+Example:
+```json
+{
+  "kind": "node",
+  "version": "v2",
+  "metadata": {
+    "name": "${AWS_ACCOUNT_ID}-${AWS_INSTANCE_ID}",
+    "labels": {
+      "env": "example"
+      "teleport.dev/discovered-node": "yes"
+    },
+  },
+  "spec": {
+    "public_addr": "ec2-54-194-252-215.us-west-1.compute.amazonaws.com",
+    "hostname": "awsxyz"
+  }
+}
+```
+
+## Agent installation
+
+In order to install the Teleport agent on EC2 instances, Teleport will serve an
+install script at `/webapi/scripts/installer`. This will be editable as a resource.
+
+Example resource script:
+```yaml
+kind: installer
```

Comment:
Is it a singleton resource? Wouldn't it be better to have multiple script names available for different instance types and specify the script per label?

Reply:
Right now it's a singleton, yes. Users can replace it with their own if they want to e.g. to detect instance type, etc. but if we wanted to have ability to have multiple scripts we could make this a non-singleton resource and for example add a field to auto-discovery configuration to optionally specify which install script to use.

```yaml
ssh_service:
  enabled: "yes"
  aws:
   - types: ["ec2"]
     regions: ["eu-central-1"]
     tags:
       "arch": "ubuntu"
     install:
       script_name: "install-script-ubuntu"
   - types: ["ec2"]
     regions: ["eu-central-1"]
     tags:
       "arch": "centos"
     install:
       script_name: "install-script-centos"
```

This would require users to mark their AWS nodes with appropriate tags if they want to use different install scripts. Auto-discovery configuration will become dynamic as a part of Teleport Discover Day-2 work so they will be update this mapping dynamically as well when we implement this.

@klizhentas @lxea What do you think?

Reply:
@lxea yes